xds/server_options.go

```diff
@@ -0,0 +1,85 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package xds
+
+import (
+	"net"
+
+	"google.golang.org/grpc"
+)
+
+// Experimental
+//
+// Notice: All APIs in this file are EXPERIMENTAL and may be changed or removed
+// in a later release.
+
+// WithServingModeCallback returns a grpc.ServerOption which allows users to
+// register a callback to get notified about serving mode changes.
+//
+// NOTE: The returned ServerOption must *only* be used in a call to
+// xds.NewGRPCServer() and must not be used in a call to grpc.NewServer().
+func WithServingModeCallback(cb ServingModeCallback) grpc.ServerOption {
+	return &smcOption{cb: cb}
+}
+
+// smcOption is a server option containing a callback to be invoked when the
+// serving mode changes.
+type smcOption struct {
+	grpc.ServerOption
+	cb ServingModeCallback
+}
+
+type serverOptions struct {
+	modeCallback ServingModeCallback
+}
+
+// ServingMode indicates the current mode of operation of the server.
+type ServingMode int
```

Comment:
Type alias to the listener package & assign these `consts` using those `consts`?

Reply:
That shouldn't happen with a type alias:

```go
type ServingMode = internalserver.ServingMode
```

Did you do it like this perhaps?
```go
type ServingMode internalserver.ServingMode
```

Reply:
Yes, I did the second one. Ok, let me try the first one. Thanks.

Reply:
Done.